application_sandbox/overlapping_frames/main.cpp

```diff
@@ -187,515 +219,460 @@ vulkan::VulkanGraphicsPipeline buildTrianglePipeline(
+  std::chrono::steady_clock::time_point start_time_point =
+      std::chrono::steady_clock::now();
+  const float triangle_speed = 0.01f;
+  GeometryPushConstantData g_push_constant_data{0.0f};
+
+  // Run the gbuffer render pass of the very first frame before entering the
+  // main loop, to initialize the interleaving work.
+  app.device()->vkResetFences(
+      app.device(), 1,
+      &frameData[current_frame].renderingFence->get_raw_object());
+
+  VkRenderPassBeginInfo g_pass_begin{
+      VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO,
+      nullptr,
+      g_render_pass,
+      g_framebuffers[current_frame].get_raw_object(),
+      {{0, 0}, {app.swapchain().width(), app.swapchain().height()}},
+      1,
+      &clear_color};
+
+  auto& cmd_buf = frameData[current_frame].gCommandBuffer;
+  auto& ref_buf = *cmd_buf;
+
+  app.BeginCommandBuffer(cmd_buf.get());
+
+  ref_buf->vkCmdBeginRenderPass(ref_buf, &g_pass_begin,
+                                VK_SUBPASS_CONTENTS_INLINE);
+  ref_buf->vkCmdBindPipeline(ref_buf, VK_PIPELINE_BIND_POINT_GRAPHICS,
+                             g_pipeline);
+  ref_buf->vkCmdPushConstants(
+      ref_buf, g_pipeline_layout, VK_SHADER_STAGE_VERTEX_BIT, 0,
+      sizeof(GeometryPushConstantData), &g_push_constant_data);
+  ref_buf->vkCmdDraw(ref_buf, 3, 1, 0, 0);
+  ref_buf->vkCmdEndRenderPass(ref_buf);
+  LOG_ASSERT(==, data->logger(), VK_SUCCESS,
+             app.EndAndSubmitCommandBuffer(
+                 cmd_buf.get(), &app.render_queue(), {}, {},
+                 {// Synchro: signal once gbuffer is terminated
+                  frameData[current_frame].gRenderFinished->get_raw_object()},
+                 VK_NULL_HANDLE));
+
+  app.device()->vkWaitForFences(app.device(), 1, &init_fence.get_raw_object(),
+                                VK_TRUE, UINT64_MAX);
+
+  // main loop
+  while (!data->WindowClosing()) {
+    // # Step 1: Prepare and submit gbuffer render pass for next_frame
+
+    // Synchro: wait on the rendering fence. This is necessary to make sure the
+    // previous postprocessing render pass on this frame index has terminated,
+    // since postprocessing consumes gbuffer results, and here we are about to
+    // edit gbuffer rendering resources.
+    app.device()->vkWaitForFences(
+        app.device(), 1,
+        &frameData[next_frame].renderingFence->get_raw_object(), VK_TRUE,
+        UINT64_MAX);
+    app.device()->vkResetFences(
+        app.device(), 1,
+        &frameData[next_frame].renderingFence->get_raw_object());
 
-    uint32_t current_frame = 0;
-    uint32_t next_frame = 1;
-    uint32_t image_index;
-    std::chrono::steady_clock::time_point start_time_point =
+    // Update push constants
+    std::chrono::steady_clock::time_point current_time_point =
         std::chrono::steady_clock::now();
-    const float triangle_speed = 0.01f;
-    GeometryPushConstantData g_push_constant_data{0.0f};
+    float time_lapse = std::chrono::duration_cast<std::chrono::microseconds>(
```

Comment:
Can we just not use a different unit here rather than microseconds in order to avoid the divide by 1000?

Reply:
Indeed! Done